File: WindowsFormsApplication6b/MainForm.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Net;
using System.Text;
using System.Text.RegularExpressions;
using System.Threading;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace WindowsFormsApplication6b
{
    public partial class MainForm : Form
    {
        private CancellationTokenSource cancellationTokenSource;
        private CancellationToken cancellationToken;

        public MainForm()
        {
            InitializeComponent();
        }

        private void getHtmlButton_Click(object sender, EventArgs e)
        {
            cancellationTokenSource = new CancellationTokenSource();
            cancellationToken = cancellationTokenSource.Token;

            var ui = TaskScheduler.FromCurrentSynchronizationContext();
            ThreadPool.QueueUserWorkItem(_ => StartRequest(ui));

            getHtmlButton.Enabled = false;
            progressBar.Enabled = true;
            progressBar.Visible = true;
            cancelButton.Enabled = true;
            cancelButton.Visible = true;
        }

        private void StartRequest(TaskScheduler ui)
        {
            byte[] inputBuffer = new byte[1024000];

            HttpWebRequest webRequest =
                (HttpWebRequest)
                WebRequest.Create("http://en.wikipedia.org/wiki/Line_of_succession_to_the_British_throne");
            webRequest.
                BeginGetResponse(asyncResult =>
                                     {
                                         WebResponse response = webRequest.EndGetResponse(asyncResult);
                                         Stream stream = response.GetResponseStream();
                                         if (stream == null) return;
                                         Task<int> bytesRead =
                                             Task<int>.Factory.
                                                 FromAsync(stream.BeginReadToEnd,
                                                           stream.EndReadToEnd,
                                                           inputBuffer, 0, inputBuffer.Length,
                                                           stream);
                                         if (this.InvokeRequired) BeginInvoke((MethodInvoker) (() => progressBar.Value = 50));
                                         bytesRead.
                                             ContinueWith(ar =>
                                                              {
                                                                  progressBar.Value = 100;
                                                                  if (!cancellationToken.IsCancellationRequested)
                                                                  {
                                                                      Trace.WriteLine(
                                                                          string.Format("Read {0} bytes",
                                                                                        ar.Result));
                                                                      string text =
                                                                          Encoding.ASCII.GetString(
                                                                              inputBuffer, 0, ar.Result);
                                                                      SetData(text);
                                                                  }
                                                                  getHtmlButton.Enabled = true;
                                                                  progressBar.Enabled = true;
                                                                  progressBar.Visible = true;
                                                                  cancelButton.Enabled = false;
                                                                  cancelButton.Visible = false;
                                                              }, ui);
                                     }, webRequest);
        }

        private void cancelButton_Click(object sender, EventArgs e)
        {
            cancellationTokenSource.Cancel();
        }

        private void SetData(string html)
        {
            listBox.Items.Clear();
            if (String.IsNullOrEmpty(html)) return;

            foreach (var x in GetScriptBodies(html))
                listBox.Items.Add(x);
        }

        private static IEnumerable<string> GetScriptBodies(string html)
        {
            const string pattern = @"<script.*?>\s*(?'scriptBody'.+?)\s*</script>";
            MatchCollection matches = Regex.Matches(html, pattern);
            if (matches.Count == 0) yield break;
            foreach (Match match in matches)
            {
                string scriptBody = match.Groups["scriptBody"].Value;
                yield return scriptBody;
            }
        }
    }
}

File: WindowsApplication6b/MainForm.vb
Imports System.Threading
Imports System.Text
Imports System.IO
Imports System.Net
Imports System.Text.RegularExpressions

Public Class MainForm
    Private _cancellationTokenSource As CancellationTokenSource
    Private _cancellationToken As CancellationToken

    Private Sub getHtmlButton_Click(sender As Object, e As EventArgs) Handles getHtmlButton.Click
        _cancellationTokenSource = New CancellationTokenSource
        _cancellationToken = _cancellationTokenSource.Token
        ThreadPool.QueueUserWorkItem(New WaitCallback(AddressOf StartRequest))

        getHtmlButton.Enabled = False
        progressBar.Enabled = True
        progressBar.Visible = True
        cancelButton.Enabled = True
        cancelButton.Visible = True
    End Sub

    Private Sub StartRequest()
        Dim inputBuffer(1024000) As Byte

        Dim request As HttpWebRequest =
                (DirectCast(WebRequest.Create("http://en.wikipedia.org/wiki/Line_of_succession_to_the_British_throne"), 
                            HttpWebRequest))

        request.
            BeginGetResponse(
                Sub(asyncResult As IAsyncResult)
                    Dim response As WebResponse = request.EndGetResponse(asyncResult)
                    Dim responseStream As Stream = response.GetResponseStream()
                    If responseStream Is Nothing Then Exit Sub

                    Dim bytesRead As Task(Of Integer) =
                            Task(Of Integer).Factory.FromAsync(AddressOf responseStream.BeginReadToEnd,
                                                                AddressOf responseStream.EndReadToEnd, inputBuffer, 0,
                                                                inputBuffer.Length, responseStream)
                    If (InvokeRequired) Then
                        BeginInvoke(Sub()
                                        progressBar.Value = 50
                                    End Sub
                                    )
                    End If
                    Dim ui As TaskScheduler = TaskScheduler.FromCurrentSynchronizationContext()
                    bytesRead.
                                ContinueWith(
                                    Sub(ar As Task(Of Integer))
                                        progressBar.Value = 100
                                        If (Not _cancellationToken.IsCancellationRequested) Then
                                            Trace.WriteLine(String.Format("Read {0} bytes", ar.Result))
                                            Dim text As String = Encoding.ASCII.GetString(inputBuffer, 0, ar.Result)
                                            Me.SetData(text)
                                        End If
                                        getHtmlButton.Enabled = True
                                        progressBar.Enabled = True
                                        progressBar.Visible = True
                                        cancelButton.Enabled = False
                                        cancelButton.Visible = False
                                        DirectCast(response, IDisposable).Dispose()
                                    End Sub, ui)
                End Sub, request)
    End Sub

    Private Sub cancelButton_Click(sender As Object, e As EventArgs) Handles cancelButton.Click
        _cancellationTokenSource.Cancel()
    End Sub

    Private Sub SetData(ByVal html As String)
        If String.IsNullOrEmpty(html) Then Return

        listBox.Items.Clear()
        For Each x As String In GetScriptBodies(html)
            listBox.Items.Add(x)
        Next
    End Sub

    Private Shared Function GetScriptBodies(ByVal html As String) As IEnumerable(Of String)
        Const pattern As String = "<script.*?>\s*(?'scriptBody'.+?)\s*</script>"
        Dim matches As MatchCollection = Regex.Matches(html, pattern)
        Dim result As List(Of String) = New List(Of String)(matches.Count)
        result.AddRange(From match As Match In matches Select match.Groups("scriptBody").Value)
        Return result
    End Function
End Class
